CSS rotate support

e3rd opened the pull request:
I came to a use-case that an image is rotated, independently to zoom.

worka commented:
I didn't understand your idea from your changes. Pls help me :)

e3rd commented:
I should have added reproduction steps before, sorry!

1. Demo https://worka.github.io/vanilla-js-wheel-zoom/examples/image.html
2. In the DevTools, lookup for `img#myContent`
3. Set CSS `rotate: 80deg`
4. Now try using the zoom: click anything on the image. Wrong point is zoomed.
5. Try dragging the image. It is being dragged in wrong directions (as if the image was not rotated, not according to the mouse move.)

My PR adds planar correction for the CSS rotate property. The rotate is being read and on zooming a point or on hovering, correct point is being zoomed / correct direction is taken.

worka commented:
https://github.com/worka/vanilla-js-wheel-zoom/issues/21 not suitable for you?

e3rd commented:
You were right to point me again to #21. My solution was not clean. I've investigated much more and found out my solution is only partial. It fixed a clash between (a) `transform: rotate()` and (b) individual `rotate`. While I start using (a), it went right. Exept the fact the wheelzoom kept meddling into the whole transform property and it was difficult to change a single subproperty be it the rotate or i.e. skew. Your solution #21 with MutationObserver works but is not so clean, adds a layer of complexity, it's more of a workaround.

In my current proposal, your excellent library stop rewriting whole transform and manages only such properties it should – translate and scale. It does not meddle to other properties like rotate. The zoom stays stable and working, whatever other properties are set to.
Moreover, it uses just the sub-parameters needed. Should the user need to flip the image (`scaleX -1`), the library now respects this. The image stays flipped.
There is no more sin nor cos used, which I agree was cubersome, left space for errors and did help with none but the rotate property itself. All the computation stays in the browser now.

worka commented:
I always try to offer solutions that do not affect the code of the library.

Firstly, it does not complicate the code of the library, and secondly, I don’t have to waste time checking and testing.

Or I give the opportunity to introduce some new behavior using callbacks.

Your solution has the right to exist, but I need time to check. Please be aware that this may take an indefinite amount of time)))

Thanks

e3rd commented:
This feature should not affect currents users. I admit it's difficult to write the tests (hence prove it) as it is a gui application. I had split screen with the old and the proposed version and tried your demo checking they behave the same way.

But there is no new feature, it's just limiting the old one. The library stops using whole transform but only individual translate and scale. It stops rewriting whole translate but only its X and Y, the same for scale. It stops using whole transition but only individual translate and scale parts. Altogether, the PR should just limit the problems people might have while using the library.

Have a nice day!

worka commented on examples/image-rotate.html:
```diff
@@ -60,28 +60,18 @@ <h1>vanilla-js-wheel-zoom</h1>
-
         rotateButton.addEventListener('click', function () {
             degreeRotation += 90;
+            degreeRotation %= 360;
```
This is a cool way to loop a value, didn't know about it before, thanks)

worka commented on src/wheel-zoom.js:
```diff
@@ -190,6 +190,10 @@ WZoom.prototype = {
 
+        const scale = content.$element.style.scale;
+        content.originalScale = scale ? scale.split(" ").map(p => parseFloat(p)) : [1,1,1];
+        content.originalTranslateZ = content.$element.style.translate?.split(" ")[2] || "0px";
```
Pls, help me understand the reason for introducing these properties( 

CSS are my weak point(

e3rd commented:
Gladly! Both scale and tr can have up to 3 parameters, the axis x, y and z.
The library uses only x and y params. When you ie set scale:4 is the
shortcut for setting scale:4 4 1 for x4 y4 and z1.

So every time you set scale:4, the z axis resets.

These lines make the library conserve original values to be used later. If
the original scale is horizontally flipped (scale:-1 1 1), it will conserve
the horizonal flip. Without these lines, it is not possible to flip and
zoom at the same time. If it is scaled on the z axis (scale:1 1 4), every
time you later set the scale, the z scale remains untouched.
The very same is for the translate Z, you translate only x and y axis and
translate z stays independent.


On Fri, Mar 29, 2024, 01:54 worka ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/wheel-zoom.js
> <https://github.com/worka/vanilla-js-wheel-zoom/pull/49#discussion_r1543909468>
> :
>
> > @@ -190,6 +190,10 @@ WZoom.prototype = {
>              content.originalHeight = options.height || content.$element.offsetHeight;
>          }
>
> +        const scale = content.$element.style.scale;
> +        content.originalScale = scale ? scale.split(" ").map(p => parseFloat(p)) : [1,1,1];
> +        content.originalTranslateZ = content.$element.style.translate?.split(" ")[2] || "0px";
>
> Pls, help me understand the reason for introducing these properties(
>
> CSS are my weak point(
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/worka/vanilla-js-wheel-zoom/pull/49#pullrequestreview-1967627275>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABU64F2XE5CXTKP3U3DFR4DY2S3VTAVCNFSM6AAAAABFHR6D7GVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSNRXGYZDOMRXGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

worka commented:
Thank you very much for the clarification. Respect. 

It became very interesting in which real case you use the library))

e3rd commented:
Here ☺ https://github.com/CZ-NIC/slidershow/ A media viewer / presenter. The library serves me to zoom everything. I came to a situation when a photo/video file was badly oriented so I needed the rotation (and might need to implement a flip in the future).

worka commented:
Hi. I accepted your PR, plz, see [branch here](https://github.com/worka/vanilla-js-wheel-zoom/tree/feature/transform). But I found bug in this example: https://worka.github.io/vanilla-js-wheel-zoom/examples/image.html. We set wrong translate/scale properties, when changing image. Help me solve this problem, i dont have free time :(

e3rd commented:
Thanks, I will!

e3rd commented:
Could you please describe more the bug? How to replicate it? I am a little confused from your description. Tried to clone the branch, but the image.html seems to work exactly like current version at https://worka.github.io/vanilla-js-wheel-zoom/examples/image.html

worka commented:
After you click on the "Change image" button, zoom starts to work incorrectly and the image moves to the side. Was I wrong?
